runtime/connectors/blob/parquetreader.go

```diff
@@ -0,0 +1,146 @@
+package blob
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/apache/arrow/go/v11/arrow"
+	"github.com/apache/arrow/go/v11/arrow/array"
+	"github.com/apache/arrow/go/v11/arrow/memory"
+	"github.com/apache/arrow/go/v11/parquet"
+	"github.com/apache/arrow/go/v11/parquet/compress"
+	"github.com/apache/arrow/go/v11/parquet/file"
+	"github.com/apache/arrow/go/v11/parquet/pqarrow"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime/pkg/arrayutil"
+	"github.com/rilldata/rill/runtime/pkg/container"
+	"gocloud.dev/blob"
+)
+
+// number of rows of a column fetched in one call
+// keeping it high seems to improve latency at the cost of accuracy in size of fetched data as per policy
+const _batchSize = int64(1024)
+
+func downloadParquet(ctx context.Context, bucket *blob.Bucket, obj *blob.ListObject, option *extractOption, fw *os.File) error {
```

Comment:
I don't have any ideas to batch this at this point since there is an intermediate step of create an arrow table for which we need to load entire data. But worth exploring from a performance point of view.

Reply:
From what I understood it is not easy to append to an existing parquet file since there is a metadata and statistics about the entire data. Most of the libraries doesn't seem to be supporting an append. 
If want to append some data that would mean reading the entire parquet file, writing data to a new file and generating the combined metadata which again will have the same memory constraints.

We could try exploring writing every row group to a separate parquet file. Max memory usage in that case will be constrained by size of a row group(with some downsides like reading multiple files and their metadata). This option can be explored in my opinion.

Some details are captured here : https://issues.apache.org/jira/browse/PARQUET-1022

Reply:
Thank you for the context – let's keep this simple and keep what we have now